### PR TITLE
[TG-2922] Improve logging for byte flatten exceptions

### PR DESCRIPTION
**Pull request by thk123**
Depends on: #2071 :white_check_mark: 

Submodule bump: https://github.com/diffblue/test-gen/pull/1739

A client has code that trips over this exception. However, we don't have access to the code so this change provides more information about the exception by re-throwing the exception back up the stack to record relevant information (like what kind of line is causing the problem). 

TODO:

- [x] Adapt existing unit test for the string utils into a catch test
- [x] Fix commit history
- [x] Add the original goto instruction corresponding to the symex step exception. 

Currently produces something like:

```
file TestClasses.java line 4 function java::Test1.f:(I)V: exception: Error converting guard for step
Souce GOTO statement: ((struct java::array[float] *)a1->data[j])->data[0] = 34.5f;
Step:
Thread 0 file TestClasses.java line 6 function java::Test1.f:(I)V bytecode-index 11
ASSIGNMENT (STATE)
dynamic_6_array#4 == (((struct java::array[float] { struct java.lang.Object { __CPROVER_string @class_identifier; _Bool @lock; } @java.lang.Object; signed int length; float *data; } *)byte_extract_little_endian(dynamic_2_array#5, POINTER_OFFSET(dynamic_object1#3.data), void *) == &dynamic_object5 ? dynamic_object5#5.data == dynamic_6_array : ((struct java::array[float] { struct java.lang.Object { __CPROVER_string @class_identifier; _Bool @lock; } @java.lang.Object; signed int length; float *data; } *)byte_extract_little_endian(dynamic_2_array#5, POINTER_OFFSET(dynamic_object1#3.data), void *) == &dynamic_object3 ? dynamic_object3#4.data == dynamic_6_array : invalid_object1#0.data == dynamic_6_array)) ? dynamic_6_array#3 WITH [0l:=34.5f] : dynamic_6_array#3)
Guard: (struct java::array[float] { struct java.lang.Object { __CPROVER_string @class_identifier; _Bool @lock; } @java.lang.Object; signed int length; float *data; } *)byte_extract_little_endian(dynamic_2_array#5, POINTER_OFFSET(dynamic_object1#3.data), void *) == &dynamic_object5 ? dynamic_object5#5.data == dynamic_6_array : ((struct java::array[float] { struct java.lang.Object { __CPROVER_string @class_identifier; _Bool @lock; } @java.lang.Object; signed int length; float *data; } *)byte_extract_little_endian(dynamic_2_array#5, POINTER_OFFSET(dynamic_object1#3.data), void *) == &dynamic_object3 ? dynamic_object3#4.data == dynamic_6_array : invalid_object1#0.data == dynamic_6_array)
 exception: Can't convert byte_extraction
  exception: cannot unpack array of non-const size
  array_type: array
    * size: symbol
        * type: signedbv
            * width: 32
            * #c_type: signed_int
        * identifier: java::Test1.f:(I)V::y!0@1#1
        * expression: symbol
            * type: signedbv
                * width: 32
                * #c_type: signed_int
            * identifier: java::Test1.f:(I)V::y
        * L0: 0
        * L1: 1
        * L2: 1
        * L1_object_identifier: java::Test1.f:(I)V::y!0@1
        * #SSA_symbol: 1
    * #dynamic: 1
    0: pointer
        * width: 64
        * #reference: 1
        0: struct
            * tag: java::array[float]
            * components: 
              0: 
                * type: struct
                    * incomplete_class: 1
                    * name: java::java.lang.Object
                    * tag: java.lang.Object
                    * base_name: java.lang.Object
                    * components: 
                      0: 
                        * type: string
                        * name: @class_identifier
                        * pretty_name: @class_identifier
                      1: 
                        * type: c_bool
                            * width: 8
                        * name: @lock
                        * pretty_name: @lock
                    * #class: 1
                * name: @java.lang.Object
                * base_name: @java.lang.Object
                * pretty_name: @java.lang.Object
              1: 
                * type: signedbv
                    * width: 32
                    * #c_type: signed_int
                * name: length
                * base_name: length
                * pretty_name: length
              2: 
                * type: pointer
                    * width: 64
                    * #reference: 1
                    0: floatbv
                        * width: 32
                        * f: 23
                * name: data
                * base_name: data
                * pretty_name: data
            * bases: 
              0: base
                  * type: symbol
                      * identifier: java::java.lang.Object
            * #class: 1
  max_bytes: +
    * type: unsignedbv
        * width: 64
        * #c_type: unsigned_long_int
    0: constant
        * type: unsignedbv
            * width: 64
            * #c_type: unsigned_long_int
        * value: 0000000000000000000000000000000000000000000000000000000000001000
    1: typecast
        * type: unsignedbv
            * width: 64
            * #c_type: unsigned_long_int
        0: pointer_offset
            * type: signedbv
                * width: 64
                * #c_type: signed_long_int
            0: member
                * type: pointer
                    * width: 64
                    * #reference: 1
                    0: pointer
                        * width: 64
                        0: empty
                * component_name: data
                0: symbol
                    * type: struct
                        * tag: java::array[reference]
                        * components: 
                          0: 
                            * type: struct
                                * incomplete_class: 1
                                * name: java::java.lang.Object
                                * tag: java.lang.Object
                                * base_name: java.lang.Object
                                * components: 
                                  0: 
                                    * type: string
                                    * name: @class_identifier
                                    * pretty_name: @class_identifier
                                  1: 
                                    * type: c_bool
                                        * width: 8
                                    * name: @lock
                                    * pretty_name: @lock
                                * #class: 1
                            * name: @java.lang.Object
                            * base_name: @java.lang.Object
                            * pretty_name: @java.lang.Object
                          1: 
                            * type: signedbv
                                * width: 32
                                * #c_type: signed_int
                            * name: length
                            * base_name: length
                            * pretty_name: length
                          2: 
                            * type: pointer
                                * width: 64
                                * #reference: 1
                                0: pointer
                                    * width: 64
                                    0: empty
                            * name: data
                            * base_name: data
                            * pretty_name: data
                        * bases: 
                          0: base
                              * type: symbol
                                  * identifier: java::java.lang.Object
                        * #class: 1
                    * identifier: symex_dynamic::dynamic_object1#3
                    * expression: symbol
                        * type: struct
                            * tag: java::array[reference]
                            * components: 
                              0: 
                                * type: struct
                                    * incomplete_class: 1
                                    * name: java::java.lang.Object
                                    * tag: java.lang.Object
                                    * base_name: java.lang.Object
                                    * components: 
                                      0: 
                                        * type: string
                                        * name: @class_identifier
                                        * pretty_name: @class_identifier
                                      1: 
                                        * type: c_bool
                                            * width: 8
                                        * name: @lock
                                        * pretty_name: @lock
                                    * #class: 1
                                * name: @java.lang.Object
                                * base_name: @java.lang.Object
                                * pretty_name: @java.lang.Object
                              1: 
                                * type: signedbv
                                    * width: 32
                                    * #c_type: signed_int
                                * name: length
                                * base_name: length
                                * pretty_name: length
                              2: 
                                * type: pointer
                                    * width: 64
                                    * #reference: 1
                                    0: pointer
                                        * width: 64
                                        0: empty
                                * name: data
                                * base_name: data
                                * pretty_name: data
                            * bases: 
                              0: base
                                  * type: symbol
                                      * identifier: java::java.lang.Object
                            * #class: 1
                        * identifier: symex_dynamic::dynamic_object1
                    * L2: 3
                    * L1_object_identifier: symex_dynamic::dynamic_object1
                    * #SSA_symbol: 1
```

Commit history still needs tidying up and I found some confusing behaviour in the string utils so want to finish off the unit tests for that.

### CODE REVIEW

**Comment by tautschnig** [comment]
I suppose that's to an extent the result of a discussion that @svorenova and I had: it is really surprising (wrong?) that Java code would generate byte_extract expressions. And actually I don't really see why this needs to be related to inaccessible client code for @svorenova has a very small example that produced this as well?!

**Comment by tautschnig** [comment]
More notes:
1. As said, the fact that Java code generates byte_extract seems broken to me. I may be wrong, but I think that's what should be fixed in the first instance.
2. This looks like pretty incredible machinery, and surely that's useful for debugging (notably C code will produce those kinds of byte_extract expressions).
3. Despite this, it's disaster recovery rather than preventing disaster. Check with @chrisr-diffblue and @danpoe on work to do the latter: such a byte_extract expression should never even hit the back-end. That's closely related to #1880.

**Comment by majakusber** [comment]
@tautschnig You are correct. Since the bug report came from a client, we are approaching this in two steps - first extending the logging for crashes of this type and second fixing the bug (by avoiding the byte extract instructions as we discussed). This is a PR for the first step, I will be preparing a PR for the second step. Extended logging should allow the client to provide us more information in the future if they encounter this crash (it took us considerable amount of time to dig out what was the actual problem).

**Comment by tautschnig** [comment]
> (it took us considerable amount of time to dig out what was the actual problem)

You are not alone in this. @chrisr-diffblue will be able to tell you more.

Edit: that's why I insist that 3. from my above comment is important.

**Comment by thk123** [comment]
@tautschnig That is mostly want I wanted to hear - that this is a reasonable approach for providing more info for these kind of errors, even when we fix the underlying problem. In this case we were able to match it up to an existing bug where we did have the source, but in general would have been hard to work out the problem without having a symmetric error in our own examples. 

I suppose there will be other errors arising from the symex phase where knowing where in the symex has been got to would be useful. This machinery will make these easier to add in in the future. 

Also thanks for the hint that @chrisr-diffblue is looking at something similar!

**Comment by thk123** [comment]
I've updated this to also print out the original goto statement that the symex has failed to convert:

```
Source GOTO statement: ((struct java::array[float] *)a1->data[j])->data[0] = 34.5f;
```

This is probably the most important piece of information (in this case I believe clearly pointing towards writing to a multi-dimensional array. 

To do this I had to thread a `ns` into it, only one level deeper but I think it is much clearer with this extra info right at the top.

**Comment by tautschnig** [comment]
> I've updated this to also print out the original goto statement that the symex has failed to convert: [...]

Very useful indeed! And then I keep wondering why this introduces a `byte_extract`...

**Review by majakusber — APPROVED** [review]
Looks good and useful. The CI failures look like minor code issues. Please consider squashing the commits for introducing/moving/adjusting the unwrap code.

**Comment by thk123** [comment]
So I split out the changes for the split_string (along with unit tests) into #2071 which this PR now depends on. Real diff for this commit (which remains changed from previous reviews): https://github.com/diffblue/cbmc/pull/2046/files/c00350b6d508ec40608a300db93d0288bd298304..705ee6592521f0682063273521a351c3244c0bf4 

Need to add the new file to the Makefile and will squash the commits as suggested by @svorenova

**Comment by thk123** [comment]
@svorenova Squashed as suggested. Also fixed the errors in Makefiles

**Comment by thk123** [comment]
Created a version that should work on Windows: https://github.com/diffblue/cbmc/pull/2046/commits/9a68db323002edb9ca13e97dc3e3b32d99e630d6

Also added a test so we actually check the logging works (fix for the bug to come)

**Comment by tautschnig** [comment]
Since the "do not merge" label has been removed I've got to ask: my thinking was that this was an experimental feature. Please do state if that's not the case and there is an actual intent to get this merged.

**Comment by thk123** [comment]
@tautschnig indeed the intention is to merge this. The goal here is to be able to provide more useful error messages for this kind of error in the future. That naturally requires the code to be in releases.

**Comment by thk123** [comment]
Windows doesn't even appear to have the `std::nested_exception` so there is no way to unroll the exception. Log a warning saying you are not getting the full picture.

**Comment by thk123** [comment]
Should have seen that coming - throw_with_nested also not supported by Windows...

**Comment by chrisr-diffblue** [comment]
Thanks @tautschnig and @thk123 for poking me about this issue. I've just had a face-to-face chat with  @thk123 about this. As @tautschnig points out, this is indeed very much a disaster recovery 'error handling of last resort' patch, but I would argue that it's still worth adding as a developer debugging aid. I don't think this is intended to be a user facing error message, and definitely we should/will be working towards ensuring that the user never hits debugging code like this where we can. @tautschnig - would you be happy to let this PR go through if there was also an appropriate issue raised to cover the 'disaster prevention' side of your concerns too?

**Comment by tautschnig** [comment]
I have stated it elsewhere, but for completeness:
>  @tautschnig - would you be happy to let this PR go through if there was also an appropriate issue raised to cover the 'disaster prevention' side of your concerns too?

I will comment on bits that I think should be changed, but certainly I don't intend to block this PR in any way.

File: src/solvers/flattening/bv_conversion_exceptions.h
Comment (tautschnig) [inline src/solvers/flattening/bv_conversion_exceptions.h]: Nit picking: can the includes be grouped by directory and be lexicographically sorted?

File: src/solvers/flattening/boolbv.cpp
Comment (tautschnig) [inline src/solvers/flattening/boolbv.cpp]: Proposed documentation: Converts an expression into its gate-level representation and returns a vector of literals corresponding to the outputs of the Boolean circuit.

File: src/goto-symex/symex_target_equation.h
Comment (tautschnig) [inline src/goto-symex/symex_target_equation.h]: This has been removed in e8eec2c8a8 so I'm a bit hesitant seeing it re-introduced. @karkhaz would you recall why it had to be removed?

File: src/goto-symex/symex_target_equation.cpp
Comment (tautschnig) [inline src/goto-symex/symex_target_equation.cpp]: I believe the same try/catch is necessary for all convert_* routines.

File: src/goto-symex/symex_target_equation.cpp
Comment (tautschnig) [inline src/goto-symex/symex_target_equation.cpp]: Don't restrict this to guards.

File: src/goto-symex/equation_conversion_exceptions.h
Comment (tautschnig) [inline src/goto-symex/equation_conversion_exceptions.h]: Use `format(...)` from `util/format_expr.h` to avoid 1) the need for a namespace and 2) language-specifics when you don't actually have information about what language it is.

File: regression/cbmc-java/dynamic-multi-dimensional-array/test.desc
Comment (tautschnig) [inline regression/cbmc-java/dynamic-multi-dimensional-array/test.desc]: The output doesn't actually include "warning: ", hence this currently fails on Windows/AppVeyor.

File: src/goto-symex/symex_target_equation.h
Comment (karkhaz) [reply src/goto-symex/symex_target_equation.h]: There's a possibility that the namespace that this reference points to would go out of scope if we save this `symex_target_equationt` and then resume it later. For some classes (e.g. `goto_symex_statet`), my solution to this was to [make the copy constructor private](https://github.com/diffblue/cbmc/blob/develop/src/goto-symex/goto_symex_state.h#L380) and [force clients to pass an equation explicitly to the copy constructor](https://github.com/diffblue/cbmc/blob/develop/src/goto-symex/goto_symex_state.h#L38).

I think the reason we decided not to do this for the `ns` member of `symex_target_equationt` was that there were no users of `ns` in the class (or maybe just the `output` method?) so we just decided to delete the member. If you would like to have a namespace reference again, I would suggest that you follow the approach I used in those links I posted above.

Comment (thk123) [reply src/goto-symex/symex_target_equation.h]: I suppose if I use `format` I can drop this back out anyway

Comment (tautschnig) [reply src/goto-symex/symex_target_equation.h]: That would be my preferred solution, yes.

Comment (thk123) [reply src/goto-symex/symex_target_equation.h]: Means we can't output the `SSA_stept` - so I'd prefer to keep it unless there is an equivalent way to work around that?

File: src/goto-symex/symex_target_equation.cpp
Comment (thk123) [inline src/goto-symex/symex_target_equation.cpp]: Copy paste error! isn't converting guard step

**Comment by thk123** [comment]
@tautschnig I've added a (very) basic version for the other conversion steps. https://github.com/diffblue/cbmc/pull/2046/commits/f7fd0a123c63a8628f4cf84036e91ae9fb060f54

I don't have the bandwidth to expand it at the moment, but I think this is at least a mild improvement for all the cases. 

I've messed up the ordering of includes in a couple of other commits, and copy pasted nonsense which I'll fix now - fixed :white_check_mark: 

Could you let me know if you're happy with:

- Keeping the `ns` for the sake of printing the `SSA_stept` 
- The extra exception stuff for the other conversion steps (modulo the message being correct!)

File: src/util/unwrap_nested_exception.cpp
Comment (thk123) [inline src/util/unwrap_nested_exception.cpp]: Include in wrong place

File: src/util/unwrap_nested_exception.h
Comment (thk123) [inline src/util/unwrap_nested_exception.h]: Includes not correctly ordered

File: src/solvers/flattening/flatten_byte_extract_exceptions.h
Comment (thk123) [inline src/solvers/flattening/flatten_byte_extract_exceptions.h]: Not ordered correctly

File: src/solvers/flattening/flatten_byte_operators.cpp
Comment (thk123) [inline src/solvers/flattening/flatten_byte_operators.cpp]: Not ordered correctly

File: src/solvers/flattening/boolbv_byte_extract.cpp
Comment (thk123) [inline src/solvers/flattening/boolbv_byte_extract.cpp]: Not ordered correctly

File: src/goto-symex/symex_target_equation.cpp
Comment (thk123) [inline src/goto-symex/symex_target_equation.cpp]: Not ordered correctly

File: src/goto-symex/symex_target_equation.cpp
Comment (thk123) [inline src/goto-symex/symex_target_equation.cpp]: Not ordered correctly

**Review by thk123 — COMMENTED** [review]
Various include orders

File: src/goto-symex/symex_target_equation.h
Comment (tautschnig) [reply src/goto-symex/symex_target_equation.h]: I'd see two options: make the `output`  function use `format` instead of `from_expr` and remove the `namespacet` parameter or introduce an `output_generic` function that does this while leaving the existing `output` function intact. In a way, such back-end code shouldn't rely on language specifics anyway and thus shouldn't be doing language-specific output either.

Comment (thk123) [reply src/goto-symex/symex_target_equation.h]: @tautschnig Created an overload without the `ns` and deprecated the old one. https://github.com/diffblue/cbmc/pull/2046/commits/069eeb1792f491a4129852ae0e947b629b4a0ee6

**Comment by thk123** [comment]
@tautschnig Due to code owner permissions, once you're happy (As implied by the :+1:), could you approve.

(@svorenova Might be worth a re-review of the final 5 commits as they are new since your review)

File: src/goto-symex/symex_target_equation.cpp
Comment (tautschnig) [inline src/goto-symex/symex_target_equation.cpp]: 1. Place all `util/` includes in one group.
2. You might add a comment with the `language_util.h` include that it can be removed once the now-deprecated function is gone.

File: src/goto-symex/symex_target_equation.h
Comment (tautschnig) [inline src/goto-symex/symex_target_equation.h:35]: I'm not currently reviewing this commit-by-commit so please excuse if my comment is wrong: this should go in a commit of its own. I agree that it can safely be removed, but this shouldn't be done as part of some random changes.

File: src/solvers/flattening/boolbv.cpp
Comment (tautschnig) [inline src/solvers/flattening/boolbv.cpp]: Please create an issue for this "TODO" and include its id in the comment text.

File: src/goto-symex/equation_conversion_exceptions.h
Comment (tautschnig) [inline src/goto-symex/equation_conversion_exceptions.h]: Not needed/used.

File: src/goto-symex/equation_conversion_exceptions.h
Comment (tautschnig) [inline src/goto-symex/equation_conversion_exceptions.h]: Not needed/used.

File: src/solvers/flattening/bv_conversion_exceptions.h
Comment (tautschnig) [inline src/solvers/flattening/bv_conversion_exceptions.h]: Neither of those seem to be needed/used.

File: src/solvers/flattening/bv_conversion_exceptions.h
Comment (tautschnig) [inline src/solvers/flattening/bv_conversion_exceptions.h]: I think only `bv_conversion_exceptions.h` is the only required one.

File: src/solvers/flattening/flatten_byte_extract_exceptions.h
Comment (tautschnig) [inline src/solvers/flattening/flatten_byte_extract_exceptions.h]: Seems unnecessary.

File: src/solvers/flattening/flatten_byte_extract_exceptions.h
Comment (tautschnig) [inline src/solvers/flattening/flatten_byte_extract_exceptions.h]: I think most of the above are unnecessary.

File: src/util/format.h
Comment (tautschnig) [inline src/util/format.h]: No, please don't. Why would this be necessary?

File: src/util/unwrap_nested_exception.cpp
Comment (tautschnig) [inline src/util/unwrap_nested_exception.cpp]: Use `"..."` within the folder instead of angle brackets.

File: src/util/unwrap_nested_exception.cpp
Comment (tautschnig) [inline src/util/unwrap_nested_exception.cpp]: Nit pick: two spaces between `in` and `a`.

File: src/util/unwrap_nested_exception.cpp
Comment (tautschnig) [inline src/util/unwrap_nested_exception.cpp]: The header file already includes those two.

File: src/util/throw_with_nested.h
Comment (majakusber) [inline src/util/throw_with_nested.h]: Typo: windowsdue to bug -> Windows due to a bug

File: src/solvers/flattening/boolbv.cpp
Comment (thk123) [reply src/solvers/flattening/boolbv.cpp]: #2103

File: src/solvers/flattening/bv_conversion_exceptions.h
Comment (thk123) [reply src/solvers/flattening/bv_conversion_exceptions.h]: These includes just seem like nonsense - `bv_conversion_exceptions.h` must not be needed since that's the file we're in! Replaced with what appear to me to be correct

File: src/util/format.h
Comment (thk123) [reply src/util/format.h]: Need the declaration where used - I thought it made more sense to include them here so people can use `format.h` without further includes, but happy to devolve this to use sites.

**Review by majakusber — APPROVED** [review]
Looks good, only one typo and please make sure to put the TODO ticket number in the code too.

File: src/goto-symex/symex_target_equation.h
Comment (thk123) [reply src/goto-symex/symex_target_equation.h:35]: So this happened as I went from `= default` -> one that takes a namespace -> `no constructor`. Will update the last commit to go back to `= default` and while I'm in the area append an extra commit to remove it. I think this is the most logical commit history

File: src/util/throw_with_nested.h
Comment (thk123) [reply src/util/throw_with_nested.h]: #2104

File: src/util/format.h
Comment (tautschnig) [reply src/util/format.h]: Yes, code should use `format_expr.h` or `format_type.h`. If `format.h` were to be used directly then this should be done consistently across the code base.

File: src/goto-symex/symex_target_equation.h
Comment (tautschnig) [reply src/goto-symex/symex_target_equation.h:35]: Brilliant, thank you very much!

**Comment by thk123** [comment]
All comments addressed I believe and TG pointer bump passing.

File: regression/cbmc-java/dynamic-multi-dimensional-array/test.desc
Comment (tautschnig) [inline regression/cbmc-java/dynamic-multi-dimensional-array/test.desc]: Nit pick: what does "This" refer to? The fact that this exception fires is some sort of bug, but the test case as set up now isn't. How about "The exception thrown in this test is the symptom of a bug; the purpose of this test is the validate the output of that exception."

File: src/goto-symex/equation_conversion_exceptions.h
Comment (tautschnig) [inline src/goto-symex/equation_conversion_exceptions.h]: `format_expr.h`

File: src/goto-symex/symex_target_equation.cpp
Comment (tautschnig) [inline src/goto-symex/symex_target_equation.cpp]: Unnecessary.

**Review by tautschnig — APPROVED** [review]
Looks good modulo the remaining nit picks.

**Pull request merged**
